apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns-from-api.php

```diff
@@ -0,0 +1,166 @@
+
+	/**
+	 * Block_Patterns constructor.
+	 */
+	private function __construct() {
+		$this->patterns_cache_key = sha1(
+			implode(
+				'_',
+				array(
+					'block_patterns',
+					PLUGIN_VERSION,
+					$this->get_iso_639_locale(),
+				)
+			)
+		);
+
+		$this->register_patterns();
+	}
+
+	/**
+	 * Creates instance.
+	 *
+	 * @return \A8C\FSE\Block_Patterns
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Register FSE block patterns and categories.
+	 */
+	private function register_patterns() {
+		if ( class_exists( 'WP_Block_Patterns_Registry' ) ) {
+			// Remove core patterns.
+			foreach ( \WP_Block_Patterns_Registry::get_instance()->get_all_registered() as $pattern ) {
+				if ( 'core/' === substr( $pattern['name'], 0, 5 ) ) {
+					unregister_block_pattern( $pattern['name'] );
+				}
+			}
+		}
+
+		$pattern_categories = array();
+		$block_patterns     = $this->get_patterns();
+
+		foreach ( (array) $this->get_patterns() as $pattern ) {
+			foreach ( (array) $pattern['categories'] as $slug => $category ) {
+				$pattern_categories[ $slug ] = $category['title'];
+			}
+		}
+
+		// Order categories alphabetically and register them.
+		asort( $pattern_categories );
+		foreach ( (array) $pattern_categories as $slug => $label ) {
+			register_block_pattern_category( $slug, array( 'label' => $label ) );
+		}
+
+		foreach ( (array) $this->get_patterns() as $pattern ) {
+			register_block_pattern(
+				Block_Patterns_From_API::PATTERN_NAMESPACE . $pattern['name'],
+				array(
+					'title'         => $pattern['title'],
+					'description'   => $pattern['title'],
+					'content'       => $pattern['html'],
+					'viewportWidth' => 1280,
+					'categories'    => array_keys(
+						$pattern['categories']
+					),
+				)
+			);
+		}
+	}
+
+	/**
+	 * Returns a list of patterns.
+	 *
+	 * @return array
+	 */
+	private function get_patterns() {
+		$block_patterns = wp_cache_get( $this->patterns_cache_key, 'ptk_patterns' );
+
+		// Load fresh data if we don't have any patterns.
+		if ( false === $block_patterns || ( defined( 'WP_DISABLE_PATTERN_CACHE' ) && WP_DISABLE_PATTERN_CACHE ) ) {
+			$request_url = esc_url_raw(
+				add_query_arg(
+					array(
+						'language' => $this->get_iso_639_locale(),
+						'tags'     => 'pattern',
+					),
+					'https://public-api.wordpress.com/rest/v1/ptk/patterns'
+				)
+			);
+
+			$args = array( 'timeout' => 20 );
+
+			if ( function_exists( 'wpcom_json_api_get' ) ) {
+				$response = wpcom_json_api_get( $request_url, $args );
+			} else {
+				$response = wp_remote_get( $request_url, $args );
+			}
+
+			if ( 200 !== wp_remote_retrieve_response_code( $response ) ) {
+				return array();
+			}
+			$block_patterns = json_decode( wp_remote_retrieve_body( $response ), true );
+			wp_cache_set( $this->patterns_cache_key, $block_patterns, 'ptk_patterns', DAY_IN_SECONDS );
```

Comment:
There was a bit of feedback in D50987-code that we could use `wp_cache_add` instead in the endpoint, I'm not sure if that applies here, too? More context for difference between the two calls in p9zhOE-Gv-p2#comment-5450

Reply:
Yes I think the same applies here.